8253750: use build-stable default seed for Utils.RANDOM_GENERATOR

iignatev opened the pull request:
Hi all,

could you please review the patch which updates `jdk.test.lib.Utils` to use md5 hash-sum of `java.vm.version` property as default seed for `Utils.RANDOM_GENERATOR`?

from JBS:
> using the same seed for all runs of a build will make it possible (easier) to compare results from different test runs (e.g. on different platforms, w/ different flags) and consequently will make test results analysis easier. the proposed solution is to use the seed based on Runtime.version() / "java.vm.version", which are different from build to build, if there is no seed specified by "jdk.test.lib.random.seed" property.

the patch also updates `RandomGeneratorTest` test, so it expects now that the same values are generated if no seed is provided. 

testing: 
✅ tier1
✅ `test/lib-test/jdk/test/lib/` against personal build on linux,windows,macos-x64
✅ `test/lib-test/jdk/test/lib/` against CI build on linux,windows,macos-x64

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8253750](https://bugs.openjdk.java.net/browse/JDK-8253750): use build-stable default seed for Utils.RANDOM_GENERATOR


### Reviewers
 * [Roger Riggs](https://openjdk.java.net/census#rriggs) (@RogerRiggs - **Reviewer**)


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/391/head:pull/391`
`$ git checkout pull/391`

iignatev commented:
/cc hotspot 
/label remove hotspot-runtime

iignatev commented:
ping?

RogerRiggs commented:
Is this really a good idea?  The purpose of using random numbers is to get broader coverage on multiple runs.
If the seed only changes once per version (6 months), that reduces test coverage.
At least for dev submitted runs, I would like to be different for every build (unless overridden).

iignatev commented:
Hi Roger,

it's exactly that you want it to be. It is different for every build (not every release, as we are using `java.vm.version` not `java.version`) unless overridden, each dev submitted run and each CI builds have different `java.vm.version` (e.g. the last two mach5 CI builds have `16-ea+19-938` and `16-ea+19-939` as their `java.vm.version`, one of my ad-hoc mach5 runs -- `16-internal+0-2020-10-01-2150482.igor.ignatyev.jdk`, my local build -- `16-internal+0-2020-10-01-2252075.iignatye...`, ) and hence would get different seeds. all the test tasks for these builds, on the other hand, would use the same seed, so one could be more confident that if test `T` passed on all platforms but platform `P`, it's platform `P` specific problem, there before this fix, one would need to rerun test `T` on each platform with the _faling_ seed and on platform `P` with at least one _passing_ seed, this problem becomes even more acute when you start considering all build flavor, vm flags, host configurations we cover in our testing.

iignatev commented:
Hi Joe,

this change indeed reduces test coverage, but not drastically (one can even argue that it doesn't reduce coverage as generally speaking you can merge coverage from runs with different configurations, e.g. run on different platforms), due to that and the reasons I wrote in my response to Roger, I don't think the reducing coverage is an issue here, nor do I think that possible impact on coverage overweights the benefits we get from comparable runs.

RogerRiggs reviewed (changes requested):
ok, if I read the code closely and know how the promoted build process works then I see your rationale.
Please update the bug report and edit the PR description to describe the conditions under which the seed
for random is computed from the build number. Its might be clearer to refer to the Runtime.Version build information
in the description.  Also make it clear that unless the system property is set, it will use a 'random' seed.

iignatev commented:
@RogerRiggs , I have updated the code to use `Runtime::version` and updated the docs to better reflect how `seed` value is being set, as well as added some explanation to both the JBS issue and PR. please let me know if it's still not clear enough.

RogerRiggs reviewed (changes requested):
The code and description are reversed from what I suggested.

The seed should *only* be computed from the version information if it is a promoted or weekly build.
All other builds should use a random seed.

iignatev commented:
Hi @RogerRiggs 

I guess I have misinterpreted your sentence about system property being set (as `java.vm.version` is always set and the only other property which mattered in this context is `jdk.test.lib.random.seed`). in any case, I agree that for _personal_ builds, it's more desirable to have different seeds on each execution, especially given the fact that the version string is set at configure-time and not at build-time, so one might end up with the same version string for a very long time.

I have reworked the code a bit, so now version-based seed is used only for _promotable_ builds (i.e. ones that have build number and it's greater than 0); local and remote/mach5 ad-hoc builds (by default) don't specify a build number, so a random seed value will be used for them. and as before, if `jdk.test.lib.random.seed` is set, its value will be used as seed oblivious to build type.

RogerRiggs reviewed (commented):
Can you take a look at the existing jdk.test.lib.RandomFactory and see if we can avoid a separate random number generator and property?

iignatev commented:
> Can you take a look at the existing jdk.test.lib.RandomFactory and see if we can avoid a separate random number generator and property?

I am not adding a new separate random number generator or a new property, `jdk.test.lib.Utils.getRandomInstance` has been around for a long time and is used by lots of tests (most of them are in `/test/hotspot/jtreg`). merging `Utils.getRandomInstance` and `RandomFactory` (or rather replacing one w/ another) is tracked by [JDK-8212077](https://bugs.openjdk.java.net/browse/JDK-8212077).

RogerRiggs commented:
I was thinking that the code added to Utils.SEED initialization could just have easily been added to RandomFactory.getSystemSeed() and not have to be changed later.

iignatev commented:
well, `RandomFactory::getSystemSeed` reads `seed` property, while `Utils.SEED` reads `jdk.test.lib.random.seed`, and there are tests which use these properties to specify seeds, I'd prefer to leave changing these tests to [8212077](https://bugs.openjdk.java.net/browse/JDK-8212077). o/c we can add a new method, e.g. `RandomFactory::getBuildStableSeed`, or refactor `RandomFactory::getSystemSeed` to take property name as an argument, but I don't think it's really worth it as 8212077 is starting to slowly bubble up in my working queue.

RogerRiggs commented:
If the factories duplication will get combined/resolved with 8212077, I'll leave that up to you.

iignatev commented:
Thanks, Roger.

/integrate
